Add option to ignore all change requests

sboardwell opened the pull request:
This PR adds the option to ignore all change requests, imitating the older `Suppress Automatic SCM trigger` option for change requests.

Our particular use-case consists of a jenkins github-multi-branch job being powered by [prow](https://github.com/kubernetes/test-infra/tree/master/prow). Currently the job is triggered twice, once from prow, and again from the webhook trigger. Disabling the webhook in github will not work because the PR's are then not detected.

sboardwell commented:
Hi @stephenc, @rsandell,

sorry for disturbing but I need a little clarification on this PR. 

Right now I'm 90% sure that this PR is not even needed. However, there may be another problem I'd like to look at. We were discussing this in the [jenkins-ci Slack channels](https://kubernetes.slack.com/archives/C09R5RQNL/p1564496772048700?thread_ts=1564086056.020300&cid=C09R5RQNL). Basically it looks like the `Skip initial build on first branch indexing` option is causing all other options to be ignored.

Could any of you help out here with a little clarification?

stephenc commented:
If you don't want change requests built at all, then don't add them to the strategy

In other words configure like:

<img width="875" alt="Screenshot 2019-07-30 at 15 38 08" src="https://user-images.githubusercontent.com/209336/62138763-1a968000-b2e0-11e9-8c16-2b8b066fe6ae.png">

Not:
<img width="873" alt="Screenshot 2019-07-30 at 15 38 18" src="https://user-images.githubusercontent.com/209336/62138792-26824200-b2e0-11e9-8ae5-226cc7ce45c9.png">

This PR is adding a checkbox to the build strategy that essentially makes it exactly the same as not adding that build strategy at all... which seems completely silly... but perhaps I am missing something subtle that you are trying to achieve... In my experience subtle things are best not included via check-boxes

stephenc commented:
The idea behind build strategies is that they just control building but do not stop discovery. An authenticated user can still manually trigger a build of a discovered branch. The discovery strategies control what branches actually get discovered in the first place

sboardwell commented:
Yes, sorry, I realise that. 

The PR here was a last ditch attempt because everything else seemed to have failed (the change requests were being triggered even though we set the Named branches strategy to only build certain branches)

> Basically it looks like the Skip initial build on first branch indexing option is causing all other options to be ignored.

Here's the message from the slack channels (the important part is found after "master was still being triggered")

Ok, I think I have it.
It seemed the Basic Build Strategy Plugin wasn’t being respected at all. It turns out that it most likely has to do with the Skip initial build on first branch indexing option.
Here’s what I know.
• My PR seems to work correctly.
• However, PR’s are still being triggered.
• I took a step back and just tested branches.
• I added some logging to the plugin to tell me when the plugin returns true.
• I added a Named branches -> Wildcard includes/excludes block with includes: my-*, excludes: master
• master was still being triggered
• Then I noticed that I also had the `Skip initial build on first branch indexing`
• I removed it and the Named branches configuration started to work as expected.
• PR’s are not triggered per default, only being triggered when the “Change Requests” option is invoked.

So all in all, it looks like the Skip initial build on first branch indexing causes all other options to be ignored, maybe because the conditional is something like `is initial, then skip, else schedule`.
I will put this into the PR and ask the Plugins owners for clarification.

sboardwell commented:
So I suppose the main question is, should setting the `Skip initial build on first branch indexing` strategy cause other options to be ignored?

stephenc commented:
> So I suppose the main question is, should setting the Skip initial build on first branch indexing strategy cause other options to be ignored?

I suspect actually a bug in that strategy's logic. I recommend closing this PR and concentrating on fixing that strategy. It should only skip the first build...

Perhaps try some debug logging to the task listener in the strategies (probably no harm to have that in all the strategies so that people can understand better what is going on)

sboardwell commented:
Will do. Thanks and sorry for the confusion 😄

sboardwell commented:
@stephenc - just a quick note to say I haven't forgotten about this and will get another PR up as soon as I can.
